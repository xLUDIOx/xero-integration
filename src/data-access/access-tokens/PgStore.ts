import { TokenSet } from 'openid-client';

import { IDbClient, INewUserTokenSetRecord, ITokenSet, IUserTokenSetRecord, SCHEMA, UserTokenSetRecordKeys } from '@shared';
import { ILogger, TenantConflictError } from '@utils';

import { IStore } from './IStore';
export class PgStore implements IStore {
    private readonly tableName: string = SCHEMA.TABLE_NAMES.ACCESS_TOKENS;

    constructor(private readonly dbClient: IDbClient, private logger: ILogger) {
    }

    async create({ account_id, user_id, tenant_id, token_set }: INewUserTokenSetRecord): Promise<void> {
        await this.preventAccountTenantConflict(account_id, tenant_id);

        await this.dbClient.query({
            text: `
                INSERT INTO "${this.tableName}" (
                    "${UserTokenSetRecordKeys.account_id}",
                    "${UserTokenSetRecordKeys.user_id}",
                    "${UserTokenSetRecordKeys.tenant_id}",
                    "${UserTokenSetRecordKeys.token_set}"
                )
                VALUES ($1, $2, $3, $4)
                ON CONFLICT ("${UserTokenSetRecordKeys.account_id}")
                DO
                    UPDATE SET
                        "${UserTokenSetRecordKeys.user_id}" = $2,
                        "${UserTokenSetRecordKeys.tenant_id}" = $3,
                        "${UserTokenSetRecordKeys.token_set}" = $4,
                        "${UserTokenSetRecordKeys.updated_at}" = now();
            `,
            values: [
                account_id,
                user_id,
                tenant_id,
                token_set,
            ],
        });
    }

    async updateToken(accountId: string, tenantId: string, tokenSet: ITokenSet): Promise<void> {
        const result = await this.dbClient.query({
            text: `
                UPDATE "${this.tableName}"
                SET
                    "${UserTokenSetRecordKeys.token_set}" = $3,
                    "${UserTokenSetRecordKeys.updated_at}" = now()
                WHERE "${UserTokenSetRecordKeys.account_id}"=$1 AND "${UserTokenSetRecordKeys.tenant_id}"=$2
                RETURNING *
            `,
            values: [
                accountId,
                tenantId,
                tokenSet,
            ],
        });

        if (result.rows.length === 0) {
            throw this.logger.child({
                accountId,
                tenantId,
                tokenSet,
            }).error(Error('Failed to update token'));
        }
    }

    async getByAccountId(accountId: string): Promise<IUserTokenSetRecord | undefined> {
        const query = await this.dbClient.query<IUserTokenSetRecord>({
            text: `
                SELECT * FROM "${this.tableName}"
                WHERE "${UserTokenSetRecordKeys.account_id}"=$1
            `,
            values: [
                accountId,
            ],
        });

        const record = query.rows[0];
        if (record) {
            record.token_set = new TokenSet(record.token_set);
        }

        return record;
    }

    async delete(tenantId: string): Promise<void> {
        await this.dbClient.query({
            text: `
                DELETE FROM "${this.tableName}"
                WHERE "${UserTokenSetRecordKeys.tenant_id}"=$1
            `,
            values: [
                tenantId,
            ],
        });
    }

    private async preventAccountTenantConflict(accountId: string, tenantId: string): Promise<void> {
        const queryResult = await this.dbClient.query<Pick<INewUserTokenSetRecord, 'account_id'>>({
            text: `
                SELECT "${UserTokenSetRecordKeys.account_id}" FROM "${this.tableName}"
                WHERE
                    "${UserTokenSetRecordKeys.account_id}"!=$1 AND
                    "${UserTokenSetRecordKeys.account_id}" NOT LIKE '%${DEMO_SUFFIX}' AND
                    "${UserTokenSetRecordKeys.tenant_id}"=$2
            `,
            values: [
                accountId,
                tenantId,
            ],
        });

        const hasConflictingAccount = queryResult.rows.length > 0;
        if (hasConflictingAccount && !accountId.endsWith(DEMO_SUFFIX)) {
            throw new TenantConflictError(tenantId, accountId, queryResult.rows[0].account_id);
        }
    }
}

const DEMO_SUFFIX = '_demo';
