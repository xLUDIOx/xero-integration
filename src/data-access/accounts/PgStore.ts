import { AccountRecordKeys, IAccountRecord, IDbClient, SCHEMA } from '@shared';

import { IStore } from './IStore';

export class PgStore implements IStore {
    private readonly tableName: string = SCHEMA.TABLE_NAMES.ACCOUNTS;

    constructor(private readonly dbClient: IDbClient) {
    }

    async get(accountId: string): Promise<IAccountRecord | undefined> {
        const query = await this.dbClient.query<IAccountRecord>({
            text: `
                SELECT * FROM "${this.tableName}"
                WHERE "${AccountRecordKeys.account_id}" = $1
            `,
            values: [accountId],
        });

        return query.rows[0];
    }

    async update(accountId: string, isSynced: boolean): Promise<void> {
        await this.dbClient.query<{ payhawk_api_key: string }>({
            text: `
                UPDATE "${this.tableName}"
                SET "${AccountRecordKeys.initial_sync_completed}"=$2
                WHERE "${AccountRecordKeys.account_id}"=$1
            `,
            values: [accountId, isSynced],
        });
    }

    async create({ account_id, tenant_id, initial_sync_completed }: IAccountRecord): Promise<void> {
        await this.dbClient.query<{ payhawk_api_key: string }>({
            text: `
                INSERT INTO "${this.tableName}" ("${AccountRecordKeys.account_id}", "${AccountRecordKeys.tenant_id}", "${AccountRecordKeys.initial_sync_completed}")
                VALUES ($1, $2, $3)
                ON CONFLICT ("${AccountRecordKeys.account_id}")
                DO NOTHING
            `,
            values: [
                account_id,
                tenant_id,
                initial_sync_completed,
            ],
        });
    }
}
