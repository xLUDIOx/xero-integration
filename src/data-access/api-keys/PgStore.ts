import { IDbClient, INewPayhawkApiKeyRecord, PayhawkApiKeyRecordKeys, SCHEMA } from '@shared';

import { IStore } from './IStore';

export class PgStore implements IStore {
    private readonly tableName: string = SCHEMA.TABLE_NAMES.PAYHAWK_API_KEYS;

    constructor(private readonly dbClient: IDbClient) {
    }

    async getByAccountId(accountId: string): Promise<string | undefined> {
        const query = await this.dbClient.query<{ key: string }>({
            text: `
                SELECT "${PayhawkApiKeyRecordKeys.key}" FROM "${this.tableName}"
                WHERE "${PayhawkApiKeyRecordKeys.account_id}" = $1
            `,
            values: [accountId],
        });

        if (query.rows.length > 0) {
            return query.rows[0].key;
        } else {
            return undefined;
        }
    }

    async set({ account_id, key }: INewPayhawkApiKeyRecord): Promise<void> {
        await this.dbClient.query<{ payhawk_api_key: string }>({
            text: `
                INSERT INTO "${this.tableName}" ("${PayhawkApiKeyRecordKeys.account_id}", "${PayhawkApiKeyRecordKeys.key}")
                VALUES ($1, $2)
                ON CONFLICT ("${PayhawkApiKeyRecordKeys.account_id}")
                DO
                    UPDATE SET "${PayhawkApiKeyRecordKeys.key}" = $2, "${PayhawkApiKeyRecordKeys.updated_at}" = NOW()
            `,
            values: [account_id, key],
        });
    }
}
