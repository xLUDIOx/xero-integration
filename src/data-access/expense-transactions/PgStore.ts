import { ExpenseTransactionRecordKeys, IDbClient, IExpenseTransactionRecord, SCHEMA } from '@shared';

import { IStore } from './IStore';

export class PgStore implements IStore {
    private readonly tableName: string = SCHEMA.TABLE_NAMES.EXPENSE_TRANSACTIONS;

    constructor(private readonly dbClient: IDbClient) {
    }

    async createIfNotExists(accountId: string, expenseId: string, transactionId: string): Promise<void> {
        await this.dbClient.query({
            text: `
                    INSERT INTO "${this.tableName}"
                        ("${ExpenseTransactionRecordKeys.account_id}", "${ExpenseTransactionRecordKeys.expense_id}", "${ExpenseTransactionRecordKeys.transaction_id}")
                    VALUES ($1, $2, $3)
                    ON CONFLICT ("${ExpenseTransactionRecordKeys.account_id}", "${ExpenseTransactionRecordKeys.expense_id}", "${ExpenseTransactionRecordKeys.transaction_id}")
                    DO NOTHING
                `,
            values: [
                accountId,
                expenseId,
                transactionId,
            ],
        });
    }

    async getByAccountId(accountId: string, expenseId: string): Promise<IExpenseTransactionRecord[]> {
        const result = await this.dbClient.query<IExpenseTransactionRecord>({
            text: `
                    SELECT * FROM "${this.tableName}"
                    WHERE "${ExpenseTransactionRecordKeys.account_id}"=$1 AND "${ExpenseTransactionRecordKeys.expense_id}"=$2
                `,
            values: [
                accountId,
                expenseId,
            ],
        });

        return result.rows;
    }

    async delete(accountId: string, expenseId: string, transactionId: string): Promise<void> {
        await this.dbClient.query<IExpenseTransactionRecord>({
            text: `
                    DELETE FROM "${this.tableName}"
                    WHERE "${ExpenseTransactionRecordKeys.account_id}"=$1 AND "${ExpenseTransactionRecordKeys.expense_id}"=$2 AND "${ExpenseTransactionRecordKeys.transaction_id}"=$3
                `,
            values: [
                accountId,
                expenseId,
                transactionId,
            ],
        });
    }
}
