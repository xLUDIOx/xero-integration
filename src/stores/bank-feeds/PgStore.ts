import { BankFeedConnectionRecordKeys, BankFeedStatementRecordKeys, IBankFeedConnectionRecord, IBankFeedStatementRecord, IDbClient, SCHEMA } from '@shared';

import { IStatementExistsFilter, IStatementFilter, IStore } from './IStore';

export class PgStore implements IStore {
    private readonly connectionsTableName: string = SCHEMA.TABLE_NAMES.BANK_FEED_CONNECTIONS;
    private readonly statementsTableName: string = SCHEMA.TABLE_NAMES.BANK_FEED_STATEMENTS;

    constructor(private readonly dbClient: IDbClient) {
    }

    async getConnectionIdByCurrency(accountId: string, currency: string): Promise<string | undefined> {
        const result = await this.dbClient.query<Pick<IBankFeedConnectionRecord, 'bank_connection_id'>>({
            text: `
                SELECT "${BankFeedConnectionRecordKeys.bank_connection_id}"
                FROM ${this.connectionsTableName}
                WHERE "${BankFeedConnectionRecordKeys.account_id}"=$1 AND
                    UPPER("${BankFeedConnectionRecordKeys.currency}")=$2
            `,
            values: [
                accountId,
                currency.toUpperCase(),
            ],
        });

        if (result.rows.length === 0) {
            return undefined;
        }

        return result.rows[0].bank_connection_id;
    }

    async getConnectionIdsForAccount(accountId: string): Promise<string[]> {
        const result = await this.dbClient.query<Pick<IBankFeedConnectionRecord, 'bank_connection_id'>>({
            text: `
                SELECT "${BankFeedConnectionRecordKeys.bank_connection_id}"
                FROM ${this.connectionsTableName}
                WHERE "${BankFeedConnectionRecordKeys.account_id}"=$1
            `,
            values: [
                accountId,
            ],
        });

        return result.rows.map(r => r.bank_connection_id);
    }

    async deleteConnectionForAccount(accountId: string, connectionId: string) {
        await this.dbClient.query({
            text: `
                DELETE FROM ${this.connectionsTableName}
                WHERE "${BankFeedConnectionRecordKeys.account_id}"=$1 AND
                    "${BankFeedConnectionRecordKeys.bank_connection_id}"=$2
            `,
            values: [
                accountId,
                connectionId,
            ],
        });
    }

    async createConnection(
        {
            account_id,
            bank_connection_id,
            currency,
        }: IBankFeedConnectionRecord,
    ): Promise<string> {
        const result = await this.dbClient.query<Pick<IBankFeedConnectionRecord, 'bank_connection_id'>>({
            text: `
                INSERT INTO ${this.connectionsTableName} (
                    "${BankFeedConnectionRecordKeys.account_id}",
                    "${BankFeedConnectionRecordKeys.bank_connection_id}",
                    "${BankFeedConnectionRecordKeys.currency}"
                )
                VALUES ($1, $2, $3)
                RETURNING "${BankFeedConnectionRecordKeys.bank_connection_id}"
            `,
            values: [
                account_id,
                bank_connection_id,
                currency,
            ],
        });

        if (result.rows.length === 0) {
            throw Error('Failed to create bank feed connection record');
        }

        return result.rows[0].bank_connection_id;
    }

    async getStatementByEntityId({ account_id, xero_entity_id, payhawk_entity_id, payhawk_entity_type }: IStatementFilter): Promise<string | undefined> {
        const whereClauses = [
            `"${BankFeedStatementRecordKeys.account_id}"=$1`,
            `"${BankFeedStatementRecordKeys.payhawk_entity_id}"=$2`,
            `"${BankFeedStatementRecordKeys.payhawk_entity_type}"=$3`,
        ];

        const values = [
            account_id,
            payhawk_entity_id,
            payhawk_entity_type,
        ];

        if (xero_entity_id) {
            whereClauses.push(`"${BankFeedStatementRecordKeys.xero_entity_id}"=$4`);
            values.push(xero_entity_id);
        }

        const result = await this.dbClient.query<Pick<IBankFeedStatementRecord, 'bank_statement_id'>>({
            text: `
                SELECT "${BankFeedStatementRecordKeys.bank_statement_id}"
                FROM ${this.statementsTableName}
                WHERE ${whereClauses.join(' AND ')}
            `,
            values,
        });

        const statementId = result.rows.length === 0 ? undefined : result.rows[0].bank_statement_id;
        return statementId;
    }

    async existsStatement({ account_id, payhawk_entity_id, payhawk_entity_type }: IStatementExistsFilter): Promise<boolean> {
        const result = await this.dbClient.query<{ exists: boolean }>({
            text: `
            SELECT EXISTS (
                SELECT 1
                FROM ${this.statementsTableName}
                WHERE "${BankFeedStatementRecordKeys.account_id}"=$1 AND
                    "${BankFeedStatementRecordKeys.payhawk_entity_id}"=$2 AND
                    "${BankFeedStatementRecordKeys.payhawk_entity_type}"=$3
            )`,
            values: [
                account_id,
                payhawk_entity_id,
                payhawk_entity_type,
            ],
        });

        return result.rows[0].exists;
    }

    async deleteStatementByEntityId({ bank_statement_id, account_id, xero_entity_id, payhawk_entity_id, payhawk_entity_type }: Required<IStatementFilter>): Promise<void> {
        await this.dbClient.query({
            text: `
                DELETE FROM ${this.statementsTableName}
                WHERE "${BankFeedStatementRecordKeys.bank_statement_id}"=$1 AND
                    "${BankFeedStatementRecordKeys.account_id}"=$2 AND
                    "${BankFeedStatementRecordKeys.xero_entity_id}"=$3 AND
                    "${BankFeedStatementRecordKeys.payhawk_entity_id}"=$4 AND
                    "${BankFeedStatementRecordKeys.payhawk_entity_type}"=$5
            `,
            values: [
                bank_statement_id,
                account_id,
                xero_entity_id,
                payhawk_entity_id,
                payhawk_entity_type,
            ],
        });
    }

    async createStatement(
        {
            account_id,
            xero_entity_id,
            payhawk_entity_id,
            payhawk_entity_type,
            bank_statement_id,
        }: IBankFeedStatementRecord,
    ): Promise<void> {
        const result = await this.dbClient.query({
            text: `
                INSERT INTO ${this.statementsTableName} (
                    "${BankFeedStatementRecordKeys.account_id}",
                    "${BankFeedStatementRecordKeys.xero_entity_id}",
                    "${BankFeedStatementRecordKeys.payhawk_entity_id}",
                    "${BankFeedStatementRecordKeys.payhawk_entity_type}",
                    "${BankFeedStatementRecordKeys.bank_statement_id}"
                )
                VALUES ($1, $2, $3, $4, $5)
                RETURNING *
            `,
            values: [
                account_id,
                xero_entity_id,
                payhawk_entity_id,
                payhawk_entity_type,
                bank_statement_id,
            ],
        });

        if (result.rows.length === 0) {
            throw Error('Failed to create bank statement record');
        }
    }
}
