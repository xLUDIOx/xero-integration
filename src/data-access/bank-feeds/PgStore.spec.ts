import { It, Mock } from 'typemoq';

import { EntityType, IDbClient } from '@shared';

import { PgStore } from './PgStore';

describe('Bank feeds store', () => {
    const dbClientMock = Mock.ofType<IDbClient>();
    const store = new PgStore(dbClientMock.object);

    beforeEach(() => {
        dbClientMock
            .setup(d => d.query(It.isAny()))
            .callback(req => expect(req).toMatchSnapshot())
            .returns(async () => ({ rows: [{}], rowCount: 1 }));
    });

    afterEach(() => {
        dbClientMock.verifyAll();
        dbClientMock.reset();
    });

    it('should match snapshot for creating feed connection', async () => {
        await store.createConnection({
            account_id: 'acc_id',
            bank_connection_id: 'conn_id',
            currency: 'BGN',
        });
    });

    it('should match snapshot for getting feed connection', async () => {
        await store.getConnectionIdByCurrency('acc_id', 'BGN');
    });

    it('should match snapshot for getting feed connection id for account', async () => {
        await store.getConnectionIdsForAccount('acc_id');
    });

    it('should match snapshot for deleting feed connections for account', async () => {
        await store.deleteConnectionForAccount('acc_id', '1232456');
    });

    it('should match snapshot for creating feed statement', async () => {
        await store.createStatement({
            account_id: 'acc_id',
            bank_statement_id: 'statement_id',
            payhawk_entity_id: 'entity_id',
            payhawk_entity_type: EntityType.Expense,
            xero_entity_id: '100',
        });
    });

    it('should match snapshot for getting feed statement', async () => {
        await store.getStatementByEntityId({
            account_id: 'acc_id',
            xero_entity_id: 'entity_id',
            payhawk_entity_id: '100',
            payhawk_entity_type: EntityType.Expense,
        });
    });

    it('should match snapshot for deleting feed statement', async () => {
        await store.deleteStatementByEntityId({
            bank_statement_id: 'statement_id',
            account_id: 'acc_id',
            xero_entity_id: 'entity_id',
            payhawk_entity_id: '100',
            payhawk_entity_type: EntityType.Expense,
        });
    });
});
