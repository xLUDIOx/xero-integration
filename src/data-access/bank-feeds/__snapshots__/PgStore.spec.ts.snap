// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bank feeds store should match snapshot for creating feed connection 1`] = `
Object {
  "text": "
                INSERT INTO bank_feed_connections (
                    \\"account_id\\",
                    \\"bank_connection_id\\",
                    \\"currency\\"
                )
                VALUES ($1, $2, $3)
                RETURNING \\"bank_connection_id\\"
            ",
  "values": Array [
    "acc_id",
    "conn_id",
    "BGN",
  ],
}
`;

exports[`Bank feeds store should match snapshot for creating feed statement 1`] = `
Object {
  "text": "
                INSERT INTO bank_feed_statements (
                    \\"account_id\\",
                    \\"xero_entity_id\\",
                    \\"payhawk_entity_id\\",
                    \\"payhawk_entity_type\\",
                    \\"bank_statement_id\\"
                )
                VALUES ($1, $2, $3, $4, $5)
                RETURNING *
            ",
  "values": Array [
    "acc_id",
    "100",
    "entity_id",
    "expense",
    "statement_id",
  ],
}
`;

exports[`Bank feeds store should match snapshot for deleting feed connections for account 1`] = `
Object {
  "text": "
                DELETE FROM bank_feed_connections
                WHERE \\"account_id\\"=$1 AND
                    \\"bank_connection_id\\"=$2
            ",
  "values": Array [
    "acc_id",
    "1232456",
  ],
}
`;

exports[`Bank feeds store should match snapshot for deleting feed statement 1`] = `
Object {
  "text": "
                DELETE FROM bank_feed_statements
                WHERE \\"bank_statement_id\\"=$1 AND
                    \\"account_id\\"=$2 AND
                    \\"xero_entity_id\\"=$3 AND
                    \\"payhawk_entity_id\\"=$4 AND
                    \\"payhawk_entity_type\\"=$5
            ",
  "values": Array [
    "statement_id",
    "acc_id",
    "entity_id",
    "100",
    "expense",
  ],
}
`;

exports[`Bank feeds store should match snapshot for getting feed connection 1`] = `
Object {
  "text": "
                SELECT \\"bank_connection_id\\"
                FROM bank_feed_connections
                WHERE \\"account_id\\"=$1 AND
                    UPPER(\\"currency\\")=$2
            ",
  "values": Array [
    "acc_id",
    "BGN",
  ],
}
`;

exports[`Bank feeds store should match snapshot for getting feed connection id for account 1`] = `
Object {
  "text": "
                SELECT \\"bank_connection_id\\"
                FROM bank_feed_connections
                WHERE \\"account_id\\"=$1
            ",
  "values": Array [
    "acc_id",
  ],
}
`;

exports[`Bank feeds store should match snapshot for getting feed statement 1`] = `
Object {
  "text": "
                SELECT \\"bank_statement_id\\"
                FROM bank_feed_statements
                WHERE \\"account_id\\"=$1 AND \\"payhawk_entity_id\\"=$2 AND \\"payhawk_entity_type\\"=$3 AND \\"xero_entity_id\\"=$4
            ",
  "values": Array [
    "acc_id",
    "100",
    "expense",
    "entity_id",
  ],
}
`;
